### DISCUSSION TITLE
Enhancement on `node`

### DISCUSSION THREAD

**Discussion by nnhjy** [spine-tools/SpineOpt.jl#801] [Ideas]
Some ideas from recent modelling practice.

## 1. `node__node` to handle distribution/allocation between different hierarchy
Suppose a system of a region which consists of several subregions having local demand, and there are both regional and subregional supply techs. (e.g. CHP for the whole region and PV for the individual subregions). It would be beneficial to have `node__node` to handle the interaction between the **region** and its **subregion**s. This can now be achieved by:
- 1. a `node` for the **region** using `connection` to link to `node`s for **subregion**s (cumbersome if users just want straightforward allocation)
- 2. a `node_group` to represent the **region** that contains `node`s for **subregion**s (complicated/impossible? if there are multi-layers)
- 3. the CHP as a `unit` directly links to the `node`s for **subregion**s via `unit__to_node` (very heavy).

The `node__node` can be used as a "shortcut" for `connection`, by which only radial topology is allowed, and for the cases where internal allocation is the concern. It can thus differentiate from using `connection` for the cases where *de facto* flows through a network are of interest. 

## 2. A node with multiple instances

Now that we have the parameter `number_of_units` for object class `unit` to represent multiple units having the same setting, it might be helpful to have it also in `node`.

### Application case
Suppose a district system model includes several building archetypes, each with the same load profile and supply tech. configuration. Each building archetype has multiple building instances within a district. Now if we want to keep the individual buildings instead of aggregating the demands, we have to set up a dedicated node for every building, making the model cumbersome. It would be nice to simply multiply the `node` as we do for the `unit`.

### Issues remain open
1. flow matching and balance between `unit`/`connection`s and `node`s. 
2. For the case above, what is actually in need is multiple instances of a group of "1 node"+"several units", so what would make more sense is that the `number_of_node` acts as a multiplier of the node and its associated units as an integrated instance. And the new abstract object_class design could be of help (#773).

Any thoughts? @jkiviluo @DillonJ @manuelma

**Comment by nnhjy** [comment]
Issue #688 covers point 1 mentioned here.

**Comment by jkiviluo** [comment]
@Tasqu is thinking about #2. There are two things in nodes that can require something like 'number_of_units'. The first one is that nodes can be storages and one needs to be able to invest in more storage. The second is the example above - it's not immediately clear to me how to distinguish these, but maybe @tasqu has some ideas. However, maybe the example in #2 is more specific. It might rather require a group that can be invested in - forcing each entity in the group to be increased by the same invest_scaling_factor (number_of_units basically).

**Comment by Tasqu** [comment]
Some comments/questions:
1. So if I understand correctly, you want to be able to point to a _region_ `node` with `units` and `connections` for convenience, while the actual output is then divided between the _subregions_? I think this is what the `node_group` is for? However, if you need some constraints limiting the allocation between the _subregions_, I'm not sure if it can be done at the moment without each _subregion_ as its own `node`. I guess further control could be beneficial, and I guess it could be controlled on the `node__node` level. However, such allocation is trespassing a bit to the domain of `connection`, and would probably require new variables etc.
2. Huh, I hadn't actually considered this quite in this way before. For my building stuff, I've only included one `node` network to represent one building archetype, and handled the `number_of_nodes` by changing the "efficiency" of the conversion unit linking it to the overarching energy network. I guess this could work, but one has to be a bit careful when scaling everything related to the archetype building, like you mention in your open issues.

In any case, I've been reformulating my archetype building approach a bit to better enable different investment options, which is kind of closer to this `number_of_nodes` approach.

**Comment by DillonJ** [comment]
Just to let you know - with the existing functionality - we use node groups to represent areas (for example Ireland and Northern Ireland) and then each node within those areas as members. The Ireland and Northern Ireland nodes have a demand time series and `balance_type=balance_type_group`. The member nodes each have a `fractional_demand` representing the fraction of the group's demand at that node. This way we can enforce ptdf or lodf based DC load flow on the lines, transport flow or disregard flows completely (copper plate) just by changing some parameters. This works quite nicely.

Not saying that we can do everything or that this is the best system - but it's good to understand what we have right now
